hail/python/hailtop/aiocloud/aioterra/azure/fs.py

```diff
@@ -0,0 +1,120 @@
+import asyncio
+import os
+from typing import AsyncContextManager, AsyncIterator, Dict, Set, Tuple, Optional
+
+from hailtop.aiocloud.aioazure import AzureAsyncFS, AzureAsyncFSURL
+from hailtop.aiotools.fs import AsyncFS, ReadableStream, WritableStream, MultiPartCreate, FileStatus, FileListEntry
+from hailtop.utils import time_msecs
+
+
+from .client import TerraClient
+
+
+WORKSPACE_STORAGE_CONTAINER_ID = os.environ.get('WORKSPACE_STORAGE_CONTAINER_ID')
+WORKSPACE_STORAGE_CONTAINER_URL = os.environ.get('WORKSPACE_STORAGE_CONTAINER_URL')
+
+
+class TerraAzureAsyncFS(AsyncFS):
+
+    def __init__(self, **azure_kwargs):
+        self._terra_client = TerraClient()
+        self._azure_fs = AzureAsyncFS(**azure_kwargs)
+        self._sas_token_cache: Dict[str, Tuple[str, int]] = {}
+
+    @property
+    def schemes(self) -> Set[str]:
+        return {'https'}
+
+    @staticmethod
+    def configured() -> bool:
+        return WORKSPACE_STORAGE_CONTAINER_ID is not None and WORKSPACE_STORAGE_CONTAINER_URL is not None
+
+    @staticmethod
+    def valid_url(url: str) -> bool:
+        return AzureAsyncFS.valid_url(url)
+
+    def parse_url(self, url: str) -> AzureAsyncFSURL:
+        return AzureAsyncFS.parse_url(url)
+
+    async def _to_azure_url(self, url: str) -> str:
+        if url in self._sas_token_cache:
+            sas_token, expiration = self._sas_token_cache[url]
+            ten_minutes_from_now = time_msecs() + 10 * 60
```

Comment:
same comment here about preferring a longer duration. I'm thinking: what if we're streaming data for a few minutes but hit a couple retries along the way?

Reply:
Do you mean that the token should live longer or we should renew the token earlier? What's here now is that we renew the token if it is within ten minutes of its expiration.

Reply:
Ya I agree this is a problem, and one that we should probably try to solve in the terra case, but I worry we won't be able to do this in the general case, possibly even in Terra. You can be given a SAS token without having the permissions to create one. They've made credential expiration an unrecoverable error. So maybe we will have to track whether we were *originally* given a SAS token or not, which should inform whether we are able to request a new one upon expiration ☹️

Reply:
Ah, sorry, I think I'm crossing some lines here between signed URLs and DRS.

Am I correct in now realizing, that your changes in the Scala-side are to support using Hail Query to read/write to the workspace storage container in particular?

---

I think there's two kinds of access patterns:

1. Batch-like. Cloud storage for relatively small objects. Object fits in RAM. Time between `open` and `close` is short (seconds).
2. Query-like. Cloud storage for analytical data. Object may not fit in RAM. Time between `open` and `close` is long (tens of minutes). Time between job specification and job execution could be >24h. Credentials must be long-lived or refreshable.

I think there's three kinds of URIs:

1. Cloud storage URLs. It has two sub-kinds.
   1. Normal cloud storage URLs. Authorization is handled by cloud IAM.
   2. Terra App workspace's container. The user will refer to this by its https:// URL. Batch service principal lacks permission to access this container. Terra Workspace Manager will mint a signed URL on demand.
3. DRS URIs. I *think* this is both an alias and authorization system. A DRS URI can decode into a cloud storage URL *and may be* a signed URL.
4. Signed URLs. These can come from decoding a DRS URI, accessing the Terra workspace manager API, or manually by a user. These have a lifetime. I'm pretty sure we cannot extend/refresh the expiration time.

---

So, it seems to me that for a Batch-like access pattern, constructing a signed URL at time-of-open is fine because the streams are always short-lived.

The expiration time is an issue for Query-like access.

>  You can be given a SAS token without having the permissions to create one.

I think my conclusion is: we must reject this. The user must provide one of: a normal cloud storage URL (which the Batch service principal is authorized to access), the Terra App's workspace container, or a DRS URI (which the Batch service principal is authorized to decode). We may also support signed URLs, but the user risks credential expiration. For the other three options:
1. Storage URLs. Easy. No changes to our code.
2. Terra App workspace container. Construct signed URL at the time of authn header creation.
3. DRS URIs. Same as (2): decode at time of authn header creation.

Reply:
And to be clear, I think for this PR we don't need to address DRS URIs. As for signed URLs, ideally we support these but, as you stated, there's nothing we can do about expiration. I think we *do* need to move URL signing to the point of authn header creation.

Reply:
Yes I agree with all of this. So in this PR the only thing we use signed URLs for are for the workspace storage container for short-lived things.

> I think we do need to move URL signing to the point of authn header creation.

Ok so we need to push down SAS token creation to any operations that issue reads/writes from the storage client. This should work so long as the storage client does not do transient error handling or issue multiple requests per client library function call…

Reply:
If we observe those issues, we can set a tight timeout on azure and hopefully handle retries ourselves.